### DISCUSSION TITLE
ArduinoCore API?

### DISCUSSION THREAD

**Discussion by bmourit** [stm32duino/Arduino_Core_STM32#2461] [General]
What are the thought of the community as far as bringing the core in line with the ArduinoCore API?

**Comment by fpistm** [comment]
Hi @bmourit 
What is your point behind this question?
Is it using the https://github.com/arduino/ArduinoCore-API ?
Most of the API are already inline with Arduino Core API and we add only some extra API.

**Reply by bmourit** [reply]
I mean it isn't following specifically. It tends to override the APIs. For example, HardwareSerial should be renamed and inherit from arduino::HardwareSerial, the ArduinoCore_API should be simlinked, etc. My only intent I guess would be more strict adherence - as in using the updated API references, following common practices found in the various official cores, and using more of the official APIs. RingBuffer, DMA, HardwareSPI, HardwareI2C, etc., where feasible. The HardwareSerial API now also defines the Serial protocols for us to use. There is also functionality added in output open drain, etc. Many things the core could use directly from the API. Would help eliminate some duplicate/custom code, while allowing faster compliance with any future Arduino API changes.

Of course this would be done slowly over time, but I don't want to start the work if it is unwanted for some specific reason. :)

Either way, the STM32duino is fantastic work and has been/is a huge support to so many in the community. I am especially in awe of the HAL and LL libraries, which was a huge effort from STM and somehow manages to give smaller, more efficient compiled code.

It would be neat to also standout as a "this is how you do an Arduino Core" and an example to other core makers of best practices.

**Reply by fpistm** [reply]
Well the example you point about `arduino::HardwareSerial` is a change made by Arduino to be able to deal with Mbed based core and is not populated to several other/older core.
I know it is feasible and like to go this way but:
* some of API have some restriction.
  * Example the SPISettings.
      https://github.com/arduino/ArduinoCore-API/blob/4a02bfc0a924e1fec34c3bb82ffd5dfba7643a0c/api/HardwareSPI.h#L40-L103
      Some month ago I've made an  update of the SPI to be more inline with the API, anyway it brought some regression and prevent to construct an SPISettings object at compiletime. 
* Activities on the repo seems very limited, several issues or PR are almost stale, so how to deal with previous example which is in fact a limitation. Could it be updated easily?
* Usage have to be done carefully versus the License. It is LGPL so we have to ensure to not add other license in the core.a.

Anyway, I welcome any contribution. As you stated it can be added slowly over time. Feel free to contribute :wink:

**Reply by bmourit** [reply]
I agree with all you said. Valid points. In regard to their older cores, they aren't listed in the Official Cores using ArduinoCore API on their readme. But like you say, perhaps they themselves are restricted in some ways.  The HardwareSerial changes may have started out as being support for Mbed, but eventually got brought together as mart of the new official API spec. To be fair, I am happy there is something finally more official, especially since on the new Arm cores came to existence.
Perhaps we start with the smaller things like the HardwareSerial, which would be pretty straightforward without much change.

**Reply by fpistm** [reply]
> * Usage have to be done carefully versus the License. It is LGPL so we have to ensure to not add other license in the core.a.

core.a is now only LGPL-2.1 since #2404.
